src/hotspot/share/code/compiledIC.cpp

```diff
@@ -75,593 +67,327 @@ bool CompiledICLocker::is_safe(address code) {
-    } else {
-      // Use icholder entry
-      assert(method_code == nullptr || method_code->is_compiled(), "must be compiled");
-      CompiledICHolder* holder = new CompiledICHolder(method(), receiver_klass);
-      info.set_icholder_entry(method()->get_c2i_unverified_entry(), holder);
-    }
-  }
-  assert(info.is_optimized() == is_optimized, "must agree");
+// GC support
+void CompiledIC::clean_metadata() {
+  data()->clean_metadata();
 }
 
+void CompiledIC::metadata_do(MetadataClosure* cl) {
+  data()->metadata_do(cl);
+}
 
-bool CompiledIC::is_icholder_entry(address entry) {
-  CodeBlob* cb = CodeCache::find_blob(entry);
-  if (cb == nullptr) {
-    return false;
-  }
-  if (cb->is_adapter_blob()) {
-    return true;
-  } else if (cb->is_vtable_blob()) {
-    return VtableStubs::is_icholder_entry(entry);
-  }
-  return false;
+#ifndef PRODUCT
+void CompiledIC::print() {
+  tty->print("Inline cache at " INTPTR_FORMAT ", calling " INTPTR_FORMAT " cached_value " INTPTR_FORMAT,
+             p2i(instruction_address()), p2i(destination()), p2i(data()));
+  tty->cr();
 }
 
-bool CompiledIC::is_icholder_call_site(virtual_call_Relocation* call_site, const CompiledMethod* cm) {
-  // This call site might have become stale so inspect it carefully.
-  address dest = cm->call_wrapper_at(call_site->addr())->destination();
-  return is_icholder_entry(dest);
+void CompiledIC::verify() {
+  _call->verify();
 }
+#endif
 
 // ----------------------------------------------------------------------------
 
-bool CompiledStaticCall::set_to_clean(bool in_use) {
+void CompiledDirectCall::set_to_clean() {
   // in_use is unused but needed to match template function in CompiledMethod
   assert(CompiledICLocker::is_safe(instruction_address()), "mt unsafe call");
   // Reset call site
-  set_destination_mt_safe(resolve_call_stub());
+  RelocIterator iter((nmethod*)nullptr, instruction_address(), instruction_address() + 1);
+  while (iter.next()) {
+    switch(iter.type()) {
+    case relocInfo::static_call_type:
+      _call->set_destination_mt_safe(SharedRuntime::get_resolve_static_call_stub());
+      break;
+    case relocInfo::opt_virtual_call_type:
+      _call->set_destination_mt_safe(SharedRuntime::get_resolve_opt_virtual_call_stub());
+      break;
+    default:
+      ShouldNotReachHere();
+    }
+  }
+  assert(is_clean(), "should be clean after cleaning");
 
-  // Do not reset stub here:  It is too expensive to call find_stub.
-  // Instead, rely on caller (nmethod::clear_inline_caches) to clear
-  // both the call and its stub.
-  return true;
+  log_debug(inlinecache)("DC@" INTPTR_FORMAT ": set to clean", p2i(_call->instruction_address()));
 }
 
-bool CompiledStaticCall::is_clean() const {
-  return destination() == resolve_call_stub();
-}
+void CompiledDirectCall::set(const methodHandle& callee_method) {
+  CompiledMethod* code = callee_method->code();
+  CompiledMethod* caller = CodeCache::find_compiled(instruction_address());
```

Comment:
Instead of doing the slow `find_compiled` unconditionally, I think it would be better to check `is_interpreted_call` first.